File: week-12/python/01_creational_patterns.py
"""
Week 12 - Creational Design Patterns in Python
"""

# 1. Singleton Pattern
class Singleton:
    """Ensures a class has only one instance"""
    
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance.value = 0
        return cls._instance
    
    def increment(self):
        self.value += 1


class DatabaseConnection(Singleton):
    """Example: Database connection as singleton"""
    
    def __init__(self):
        if not hasattr(self, 'initialized'):
            self.host = "localhost"
            self.port = 5432
            self.initialized = True
    
    def connect(self):
        return f"Connected to {self.host}:{self.port}"


# 2. Factory Pattern
class Animal:
    """Base animal class"""
    
    def speak(self):
        pass


class Dog(Animal):
    def speak(self):
        return "Woof!"


class Cat(Animal):
    def speak(self):
        return "Meow!"


class AnimalFactory:
    """Factory to create animals"""
    
    @staticmethod
    def create_animal(animal_type):
        if animal_type == "dog":
            return Dog()
        elif animal_type == "cat":
            return Cat()
        else:
            raise ValueError(f"Unknown animal type: {animal_type}")


# 3. Builder Pattern
class Pizza:
    """Product class"""
    
    def __init__(self):
        self.size = None
        self.crust = None
        self.toppings = []
    
    def __str__(self):
        return f"{self.size} pizza with {self.crust} crust and toppings: {', '.join(self.toppings)}"


class PizzaBuilder:
    """Builder class"""
    
    def __init__(self):
        self.pizza = Pizza()
    
    def set_size(self, size):
        self.pizza.size = size
        return self
    
    def set_crust(self, crust):
        self.pizza.crust = crust
        return self
    
    def add_topping(self, topping):
        self.pizza.toppings.append(topping)
        return self
    
    def build(self):
        return self.pizza


if __name__ == "__main__":
    print("=" * 60)
    print("Singleton Pattern")
    print("=" * 60)
    
    s1 = Singleton()
    s2 = Singleton()
    
    print(f"s1 is s2: {s1 is s2}")  # True - same instance
    
    s1.increment()
    print(f"s1.value: {s1.value}")
    print(f"s2.value: {s2.value}")  # Same value
    
    db1 = DatabaseConnection()
    db2 = DatabaseConnection()
    print(f"\ndb1 is db2: {db1 is db2}")
    print(db1.connect())
    
    print("\n" + "=" * 60)
    print("Factory Pattern")
    print("=" * 60)
    
    dog = AnimalFactory.create_animal("dog")
    cat = AnimalFactory.create_animal("cat")
    
    print(f"Dog says: {dog.speak()}")
    print(f"Cat says: {cat.speak()}")
    
    print("\n" + "=" * 60)
    print("Builder Pattern")
    print("=" * 60)
    
    pizza = (PizzaBuilder()
            .set_size("Large")
            .set_crust("Thin")
            .add_topping("Cheese")
            .add_topping("Pepperoni")
            .add_topping("Mushrooms")
            .build())
    
    print(pizza)

File: week-12/cpp/01_creational_patterns.cpp
/**
 * Week 12 - Creational Design Patterns in C++
 */

#include <iostream>
#include <string>
#include <vector>
#include <memory>

using namespace std;

// 1. Singleton Pattern
class Singleton {
private:
    static Singleton* instance;
    int value;
    
    // Private constructor
    Singleton() : value(0) {}

public:
    // Delete copy constructor and assignment operator
    Singleton(const Singleton&) = delete;
    Singleton& operator=(const Singleton&) = delete;
    
    static Singleton* getInstance() {
        if (instance == nullptr) {
            instance = new Singleton();
        }
        return instance;
    }
    
    void increment() {
        value++;
    }
    
    int getValue() const {
        return value;
    }
};

// Initialize static member
Singleton* Singleton::instance = nullptr;


// 2. Factory Pattern
class Animal {
public:
    virtual string speak() = 0;
    virtual ~Animal() {}
};

class Dog : public Animal {
public:
    string speak() override {
        return "Woof!";
    }
};

class Cat : public Animal {
public:
    string speak() override {
        return "Meow!";
    }
};

class AnimalFactory {
public:
    static unique_ptr<Animal> createAnimal(const string& type) {
        if (type == "dog") {
            return make_unique<Dog>();
        } else if (type == "cat") {
            return make_unique<Cat>();
        }
        return nullptr;
    }
};


// 3. Builder Pattern
class Pizza {
private:
    string size;
    string crust;
    vector<string> toppings;

public:
    void setSize(const string& s) { size = s; }
    void setCrust(const string& c) { crust = c; }
    void addTopping(const string& t) { toppings.push_back(t); }
    
    void display() const {
        cout << size << " pizza with " << crust << " crust and toppings: ";
        for (size_t i = 0; i < toppings.size(); i++) {
            cout << toppings[i];
            if (i < toppings.size() - 1) cout << ", ";
        }
        cout << endl;
    }
};

class PizzaBuilder {
private:
    Pizza pizza;

public:
    PizzaBuilder& setSize(const string& size) {
        pizza.setSize(size);
        return *this;
    }
    
    PizzaBuilder& setCrust(const string& crust) {
        pizza.setCrust(crust);
        return *this;
    }
    
    PizzaBuilder& addTopping(const string& topping) {
        pizza.addTopping(topping);
        return *this;
    }
    
    Pizza build() {
        return pizza;
    }
};


int main() {
    cout << string(60, '=') << endl;
    cout << "Singleton Pattern" << endl;
    cout << string(60, '=') << endl;
    
    Singleton* s1 = Singleton::getInstance();
    Singleton* s2 = Singleton::getInstance();
    
    cout << "s1 == s2: " << (s1 == s2 ? "true" : "false") << endl;
    
    s1->increment();
    cout << "s1 value: " << s1->getValue() << endl;
    cout << "s2 value: " << s2->getValue() << endl;
    
    cout << "\n" << string(60, '=') << endl;
    cout << "Factory Pattern" << endl;
    cout << string(60, '=') << endl;
    
    auto dog = AnimalFactory::createAnimal("dog");
    auto cat = AnimalFactory::createAnimal("cat");
    
    cout << "Dog says: " << dog->speak() << endl;
    cout << "Cat says: " << cat->speak() << endl;
    
    cout << "\n" << string(60, '=') << endl;
    cout << "Builder Pattern" << endl;
    cout << string(60, '=') << endl;
    
    Pizza pizza = PizzaBuilder()
                    .setSize("Large")
                    .setCrust("Thin")
                    .addTopping("Cheese")
                    .addTopping("Pepperoni")
                    .addTopping("Mushrooms")
                    .build();
    
    pizza.display();
    
    return 0;
}
